### PR TITLE
Introduce LLVMCPUVectorLowering to get rid of StrategyLowerVectors pass.

### PR DESCRIPTION
**Pull request by hanhanW**
It refactors seven stages into three stages. The pattern set of first two stages is inspired by LLVMGPUVectorLowering, and the last stage is for vector.transpose.

### CODE REVIEW

**Comment by hanhanW** [comment]
There are two commits in the PR. Hopefully it's easier to review.

1. Adds the pass and uses it in CPU pipeline. https://github.com/openxla/iree/pull/12845/commits/491957e44c5957af856a660bec413f00fb972879
2. Delete the implementation of codegen strategy. It only contains code deletion. https://github.com/openxla/iree/pull/12845/commits/3984299164fb06d6775bfad53e93510b8458e0e5

**Review by dcaballe — CHANGES_REQUESTED** [review]
One of the main pain points of vector lowering is that there are too many patterns being applied at the same time with no visibility between each rewrite. I would take this cleanup opportunity as an opportunity to also address that problem and to move as many patterns as possible to independent passes. At least, I would create independent passes to lower transposes, reductions, contractions, transfer split (?, what is a transfer split?), shape casts and transfer ops. The vector to SCF conversion patterns should be independent as well, IMO. WDYT?

Thanks a lot for doing this!

**Comment by MaheshRavishankar** [comment]
> One of the main pain points of vector lowering is that there are too many patterns being applied at the same time with no visibility between each rewrite. I would take this cleanup opportunity as an opportunity to also address that problem and to move as many patterns as possible to independent passes. At least, I would create independent passes to lower transposes, reductions, contractions, transfer split (?, what is a transfer split?), shape casts and transfer ops. The vector to SCF conversion patterns should be independent as well, IMO. WDYT?
> 
> Thanks a lot for doing this!

I am not sure I would add them as separate passes. That seems like an artificial boundary. I'd expect the pass to vectorize as many operations as it can. If there is an ordering constraint that should be part of the pass itself. We can have multiple fixed points in a single pass if needed

**Comment by dcaballe** [comment]
Otherwise, we would be preserving the main blackbox of codegen driver and that wouldn't help debugging (i.e., `-mlir-print-ir-after-all`), which, to reiterate, has been a main pain point. This pass is doing far too many random things... from lowering contractions using 4 or 5 different strategies, to lowering some transposes to x86 asm, applying a large amount of canonicalization patterns, canonicalizing xfer ops, lowering operations to SCF... it's far to much. If we don't want to create unitary passes, let's try to find small groups of patterns that make sense to apply together. Does that sound more reasonable?

**Comment by dcaballe** [comment]
A similar example is the pass that lowers MLIR to the LLVM dialect. That pass has been an issue for debugging as it's doing far too many things in one shot... It's even lowering some TOSA operations...

**Comment by MaheshRavishankar** [comment]
> Otherwise, we would be preserving the main blackbox of codegen driver and that wouldn't help debugging (i.e., `-mlir-print-ir-after-all`), which, to reiterate, has been a main pain point. This pass is doing far too many random things... from lowering contractions using 4 or 5 different strategies, to lowering some transposes to x86 asm, applying a large amount of canonicalization patterns, canonicalizing xfer ops, lowering operations to SCF... it's far to much. If we don't want to create unitary passes, let's try to find small groups of patterns that make sense to apply together. Does that sound more reasonable?

I don't disagree with that. We could basically add IR dumps behind a debug flag. Not print IR after all, but still can be used. 
I am not overly concerned about having them as different passes (though that seems like an artificial demarcation for just printing purposes). I am more concerned about landing this soon and deleting Codegen driver.

**Comment by MaheshRavishankar** [comment]
> A similar example is the pass that lowers MLIR to the LLVM dialect. That pass has been an issue for debugging as it's doing far too many things in one shot... It's even lowering some TOSA operations...

It's a tangent, but I prefer it the way it is now. I think dialect conversion works best if you just lower everything one shot. Splitting them again seems like an artificial break and having passes that have to go together. But if someone wants to break it up, that'd be fine too..

**Comment by dcaballe** [comment]
It's not only about the IR dumps but all the pass manager features (IR dumps, debug flags, filters, timing...). I don't think we should reimplement any of this features within a pass. Let's discuss this over GVC to see if we can find a middle ground.

**Comment by MaheshRavishankar** [comment]
> It's not only about the IR dumps but all the pass manager features (IR dumps, debug flags, filters, timing...). I don't think we should reimplement any of this features within a pass. Let's discuss this over GVC to see if we can find a middle ground.

Yeah, happy to chat over GVC, but I am not overly opinionated about this. Whatever you guys agree on works for me (I can help tie breaking if you want to break ties, but probably not needed). Like I said, my main concern is related to landing this and deleting a whole bunch of legacy code.

**Comment by hanhanW** [comment]
> One of the main pain points of vector lowering is that there are too many patterns being applied at the same time with no visibility between each rewrite. I would take this cleanup opportunity as an opportunity to also address that problem and to move as many patterns as possible to independent passes. At least, I would create independent passes to lower transposes, reductions, contractions, transfer split (?, what is a transfer split?), shape casts and transfer ops. The vector to SCF conversion patterns should be independent as well, IMO. WDYT?

I understand that it's a pain point about having a lot of patterns in a lowering. (I've been suffered from that too) Before the deprecation, there are seven "passes" and each of them is a super set of the previous one. I thought that I've addressed the issue a alot, though it's still many in the first stage. I've spent a lot of time on breaking things down. During the refactoring, I realized that the order matters. It results in different IR if we swap some stages. Thus, I keep them in the same pass. We can add some debug messages to dump the IR.

I'm not a vector lowering expert; I'm welcome to layer things properly. This is an acceptable snapshot to me because

1. We reduce seven big passes to "three stages in the new pass". Besides canonicalization patterns, the patterns are not duplicated in different stages.
2. This is what CUDA does at this moment. Not saying that this is a more proper way, but it seems a reasonable way to me.

The goal of the PR is deprecating CodegenStrategy and simplifying vector lowering pass. If we have concrete picture about how things can be separated, I'm happy to structure the passes in that way. Otherwise, I purpose to keep the structure as it is, land the PR, and leave the better structure as a follow up.

Note: My observation is that some patterns must be run together in first stage. I can't tell that what they are because there are too many combination and my experiment was 1~2 weeks ago..

> This pass is doing far too many random things... from lowering contractions using 4 or 5 different strategies, to lowering some transposes to x86 asm, applying a large amount of canonicalization patterns, canonicalizing xfer ops, lowering operations to SCF.

The existing CPU pipeline only care `splitVectorTransfersTo` option and `lowerVectorTransposeToAVX2` option. I thought that we'd like to keep the control about strategies, but it seems not. I can delete all unused options, so we will just have two options.

**Comment by MaheshRavishankar** [comment]
> The existing CPU pipeline only care `splitVectorTransfersTo` option and `lowerVectorTransposeToAVX2` option. I thought that we'd like to keep the control about strategies, but it seems not. I can delete all unused options, so we will just have two options.

That might be worth it. We cleanup things that we dont need, and then we explcitly add what we need.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUVectorLowering.cpp [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        patterns, vectorTransferToSCFOptions.setTargetRank(maxTransferRank));
+    (void)applyPatternsAndFoldGreedily(getOperation(), std::move(patterns));
+  }
+
+  {
+    // Lowering for vector.transpose ops.
+    RewritePatternSet patterns(&getContext());
+    vector::populateVectorToVectorCanonicalizationPatterns(patterns);
+    vector::populateVectorTransposeLoweringPatterns(patterns,
+                                                    vectorTransformOptions);
```
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUVectorLowering.cpp:98]: Where is this vectorTransformOptions declared?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUVectorLowering.cpp [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            .Case("none", vector::VectorTransferSplit::None)
+            .Case("linalg-copy", vector::VectorTransferSplit::LinalgCopy)
+            .Case("vector-transfers",
+                  vector::VectorTransferSplit::VectorTransfer)
+            .Default(vector::VectorTransferSplit::None);
+
+    // Per-function lowering pipeline.
+    vector::VectorTransformsOptions vectorTransformOptions =
+        vector::VectorTransformsOptions()
+            .setVectorTransposeLowering(vectorTransposeLowering)
```
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUVectorLowering.cpp]: transpose options shouldn't be necessary here as we are not lowering them in this step.

**Review by dcaballe — APPROVED** [review]
Fair enough. Let's move forward with this and revisit later. If we don't want to split by op, perhaps we can do some sort of VectorToVector, VectorToSCF and other specialized patterns. We can discuss this offline.

**Pull request merged**